edgedb/_testbase.py

```diff
@@ -52,52 +54,76 @@ def flt(log_record):
 _default_cluster = None
 
 
-def _init_cluster(data_dir=None, *, cleanup_atexit=True):
-    if (not os.environ.get('EDGEDB_DEBUG_SERVER') and
-            not os.environ.get('EDGEDB_LOG_LEVEL')):
-        _env = {'EDGEDB_LOG_LEVEL': 'silent'}
-    else:
-        _env = {}
-
-    if data_dir is None:
-        cluster = edgedb_cluster.TempCluster(env=_env, testmode=True)
-        destroy = True
-    else:
-        cluster = edgedb_cluster.Cluster(data_dir=data_dir, env=_env)
-        destroy = False
-
-    if cluster.get_status() == 'not-initialized':
-        cluster.init()
-
-    cluster.start(port='dynamic')
-    cluster.set_superuser_password('test')
-
-    if cleanup_atexit:
-        atexit.register(_shutdown_cluster, cluster, destroy=destroy)
-
-    return cluster
-
-
 def _start_cluster(*, cleanup_atexit=True):
     global _default_cluster
 
-    if _default_cluster is None:
-        cluster_addr = os.environ.get('EDGEDB_TEST_CLUSTER_ADDR')
-        if cluster_addr:
-            conn_spec = json.loads(cluster_addr)
-            _default_cluster = edgedb_cluster.RunningCluster(**conn_spec)
+    if isinstance(_default_cluster, Exception):
+        # when starting a server fails
+        # don't retry starting one for every TestCase
+        # because repeating the failure can take a while
+        raise _default_cluster
+
+    if _default_cluster:
+        return _default_cluster
+
+    try:
+        tmpdir = tempfile.TemporaryDirectory()
+        status_file = os.path.join(tmpdir.name, 'server-status')
+
+        # if running on windows adjust the path for WSL
+        status_file_unix = (
+            status_file.replace('C:', '/mnt/c')
+            .replace("\\", '/')
+            .lower()
+        )
+
+        args = [
+            os.environ.get('EDGEDB_SERVER_BINARY', 'edgedb-server'),
+            "--temp-dir",
+            "--testmode",
+            f"--emit-server-status={status_file_unix}",
+            "--port=auto",
+            "--bootstrap-command=ALTER ROLE edgedb { SET password := 'test' }",
```

Comment:
Is there a reason why there isn't `--auto-shutdown` here?

Reply:
The test suite currently creates a connection per `TestCase` class. With `--auto-shutdown` the server exits after the first `TestCase` closes it's connection and the rest of the tests fail.

Reply:
Hmm. I don't see how the server is being shut down then.  I think we should modify `--auto-shutdown` to accept a timeout.

Reply:
I added a global connection and `atexit.register(con.close)` so that there is a persistent connection throughout the test run.

Reply:
and added `--auto-shutdown`

Reply:
OK. Still, let's implement shutdown delay in the server.  I think something like 5 seconds by default should address the issue and we can make it configurable via a new `--auto-shutdown-after=Ns` option or something like that.

Reply:
https://github.com/edgedb/edgedb/issues/2565